### DISCUSSION TITLE
rewrites: Allow rewrites to be applied based on the request method

### DISCUSSION THREAD

**Discussion by Vivalldi** [vercel/next.js#29304] [Ideas]
#### How I use Next.js
Next.js generates our static assets (`next build && next export`). Files are then served by a Go server. API routes & forms are handled by the Go server.

#### What I want
I want to be able to use `next dev` to run the server in development. I want to rewrite form requests to the Go server (at a different address than next). I want these form requests to be at the same path; `GET /sign-in` serves a Next.js page, `POST /sign-in` is rewritten to the backend server.

### Describe the solution you'd like

Add a `method` type to the ["has objects"](https://nextjs.org/docs/api-reference/next.config.js/rewrites#header-cookie-and-query-matching)

```js
module.exports = {
  async rewrites() {
    return [
      // If method is "POST" this rewrite will be applied
      {
        source: "/:path*",
        has: [
            { type: "method", value: "POST" }
        ],
        destination: "http://other-dev-server:8000/:path*",
      }
    ];
  },
};
```

#### Alternative Feature - Output dev to static
Run `next build` & `next export` in a watchable manner

#### Workaround - Serve assets during dev from Go
One workaround to this is to always serve the assets from the go server. This comes at the cost of needing to run `next build && next export` for each change made as well as restarting the go server if you're embedding assets (we serve from the file system during dev)

#### Workaround - Alternative paths
The best alternative right now is to specify a different URL path entirely (`/api/...`) to distinguish between server requests vs "asset" requests. This can be done trivially using the existing rewrites functionality. This allows us to develop using the `next dev` server to maintain watch & auto-refresh functionalities. It just means we have to account for the various routes slightly differently in Go 

```js
module.exports = {
  async rewrites() {
    return [
      {
        source: "/api/:path*",
        destination: "http://other-dev-server:8000/api/:path*",
      }
    ];
  },
};

```

**Comment by Vivalldi** [comment]
(Apologies for creating an issue, didn't realize we had moved to discussions)

The best workaround I've found so far: rewrite based on `content-type` header. It's not perfect, but it gets the job done

```js
module.exports = {
  async rewrites() {
    return {
      beforeFiles: [
        // These rewrites are checked after headers/redirects
        // and before all files including _next/public files which
        // allows overriding page files
        {
          source: "/:path*",
          destination: "http://localhost:9100/:path*",
          has: [{ type: "header", kype: "header", key: "content-type" }],
        },
      ],
      fallback: [
        // These rewrites are checked after both pages/public files
        // and dynamic routes are checked
        {
          source: "/:path*",
          destination: `http://localhost:9100/:path*`,
        },
      ],
    };
  },
};
```

**Comment by cbrianball** [comment]
I have implemented this in my fork:
https://github.com/cbrianball/next.js/tree/routeHas-HttpMethod

I went to fill out the PR, but there's a spot to check that it's been approved for implementation, and it doesn't look like this has.

Also, I've only tested this locally and understand that I would need to add tests and modify examples & docs before it's fully done, but I wanted to get initial feedback first to make sure this is something that may actually be accepted, or if there's an underlying reason why this isn't a "good idea".

Thanks!

**Comment by icyJoseph** [comment]
I think, today, with Next 12, the best way forward for this is to do `_middleware` at the top of the pages tree.

```ts
import { NextRequest, NextResponse } from "next/server";

export const middleware = (req: NextRequest) => {
  const { pathname } = req.nextUrl;
  
  if (req.method === "POST"  && pathname === '/no-post') {
   // I guess you want to check for more properties before doing this
    return NextResponse.rewrite("http://example.com");
  }

  return NextResponse.next();
};
```

**Reply by cbrianball** [reply]
That makes sense. Thanks!
